tokio-sync/src/mpsc/chan.rs

```diff
@@ -0,0 +1,394 @@
+where
+    S: Semaphore,
+{
+    fn drop(&mut self) {
+        use super::block::Read::Value;
+
+        self.close();
+
+        let rx_fields = unsafe { &mut *self.inner.rx_fields.get() };
+
+        while let Some(Value(_)) = rx_fields.list.pop(&self.inner.tx) {
+            self.inner.semaphore.add_permit();
+        }
+    }
+}
+
+// ===== impl Chan =====
+
+impl<T, S> Drop for Chan<T, S> {
+    fn drop(&mut self) {
+        use super::block::Read::Value;
+
+        let rx_fields = unsafe { &mut *self.rx_fields.get() };
+
+        while let Some(Value(_)) = rx_fields.list.pop(&self.tx) {
+        }
+    }
+}
+
+use semaphore::TryAcquireError;
+
+impl From<TryAcquireError> for TrySendError {
+    fn from(src: TryAcquireError) -> TrySendError {
+        if src.is_closed() {
+            TrySendError::Closed
+        } else if src.is_no_permits() {
+            TrySendError::NoPermits
+        } else {
+            unreachable!();
+        }
+    }
+}
+
+// ===== impl Semaphore for (::Semaphore, capacity) =====
+
+use semaphore::Permit;
+
+impl Semaphore for (::semaphore::Semaphore, usize) {
+    type Permit = Permit;
+
+    fn new_permit() -> Permit {
+        Permit::new()
+    }
+
+    fn drop_permit(&self, permit: &mut Permit) {
+        if permit.is_acquired() {
+            permit.release(&self.0);
+        }
+    }
+
+    fn add_permit(&self) {
+        self.0.add_permits(1)
+    }
+
+    fn is_idle(&self) -> bool {
+        self.0.available_permits() == self.1
+    }
+
+    fn poll_acquire(&self, permit: &mut Permit) -> Poll<(), ()> {
+        permit.poll_acquire(&self.0)
+            .map_err(|_| ())
+    }
+
+    fn try_acquire(&self, permit: &mut Permit) -> Result<(), TrySendError> {
+        permit.try_acquire(&self.0)?;
+        Ok(())
+    }
+
+    fn forget(&self, permit: &mut Self::Permit) {
+        permit.forget()
+    }
+
+    fn close(&self) {
+        self.0.close();
+    }
+}
+
+// ===== impl Semaphore for AtomicUsize =====
+
+use std::sync::atomic::Ordering::{Acquire, Release};
+use std::usize;
+
+impl Semaphore for AtomicUsize {
+    type Permit = ();
+
+    fn new_permit() {
+    }
+
+    fn drop_permit(&self, _permit: &mut ()) {
+    }
+
+    fn add_permit(&self) {
+        let prev = self.fetch_sub(2, Release);
+
+        if prev >> 1 == 0 {
+            // Something went wrong
+            process::abort();
+        }
+    }
+
+    fn is_idle(&self) -> bool {
+        self.load(Acquire) >> 1 == 0
+    }
+
+    fn poll_acquire(&self, permit: &mut ()) -> Poll<(), ()> {
+        use futures::Async::Ready;
+        self.try_acquire(permit)
+            .map(Ready)
+            .map_err(|_| ())
+    }
+
+    fn try_acquire(&self, _permit: &mut ()) -> Result<(), TrySendError> {
+        let mut curr = self.load(Acquire);
+
+        loop {
+            if curr & 1 == 1 {
+                return Err(TrySendError::Closed);
+            }
+
+            if curr == usize::MAX ^ 1 {
+                process::abort()
```

Comment:
I assume it's similar to `Arc`, where cloning too many times risks overflowing the ref count, and `Drop` would free memory too early, so it's not safe to just panic.

Reply:
Yes, what @seanmonstar said. Also, it would be pretty hard to hit this case without running up to process memory limits. It is more for theoretical correctness.